File: java/src/edu/caltech/ipac/firefly/visualize/WebPlotInitializer.java
/*
 * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
 */
package edu.caltech.ipac.firefly.visualize;

import edu.caltech.ipac.firefly.data.DataEntry;
import edu.caltech.ipac.visualize.plot.CoordinateSys;
import edu.caltech.ipac.visualize.plot.projection.Projection;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
/**
 * User: roby
 * Date: Sep 11, 2009
 * Time: 11:16:02 AM
 */


/**
 * @author Trey Roby
 */
public class WebPlotInitializer implements Serializable, DataEntry {

    private final static String SPLIT_TOKEN= "--WebPlotInitializer--";

    private CoordinateSys _imageCoordSys;
    private String    _projectionSerialized;
    private int           _dataWidth;
    private int           _dataHeight;
    private int           _imageScaleFactor;
    private PlotImages    _initImages;
    private PlotState     _plotState;
    private WebFitsData   _fitsData[];
    private String        _desc;
    private String        _dataDesc;



    private WebPlotInitializer() {}


    public WebPlotInitializer(PlotState plotState,
                             PlotImages images,
                             CoordinateSys imageCoordSys,
                             Projection projection,
                             int dataWidth,
                             int dataHeight,
                             int imageScaleFactor,
                             WebFitsData  fitsData[],
                             String desc,
                             String dataDesc) {

        _plotState= plotState;
        _initImages= images;
        _imageCoordSys= imageCoordSys;
        _projectionSerialized= ProjectionSerializer.serializeProjection(projection);
        _dataWidth= dataWidth;
        _dataHeight= dataHeight;
        _imageScaleFactor= imageScaleFactor;
        _fitsData= fitsData;
        _desc= desc;
        _dataDesc= dataDesc;
    }

//======================================================================
//----------------------- Public Methods -------------------------------
//======================================================================
    
    public PlotState getPlotState() { return _plotState; }
    public CoordinateSys getCoordinatesOfPlot() { return _imageCoordSys; }
    public PlotImages getInitImages() { return _initImages; }

    public Projection getProjection() {
//        return _projection;
        return ProjectionSerializer.deserializeProjection(_projectionSerialized);
    }

    public int getDataWidth() { return _dataWidth; }
    public int getDataHeight() { return _dataHeight; }
    public int getImageScaleFactor() { return _imageScaleFactor; }
    public WebFitsData[] getFitsData()  { return _fitsData; }

    public String getPlotDesc() { return _desc; }
    public void setPlotDesc(String d) { _desc= d; }
    public String getDataDesc() { return _dataDesc; }

    public String toString() {
        StringBuilder sb= new StringBuilder(300);
        sb.append(_imageCoordSys).append(SPLIT_TOKEN);
        sb.append(_projectionSerialized).append(SPLIT_TOKEN);
        sb.append(_dataWidth).append(SPLIT_TOKEN);
        sb.append(_dataHeight).append(SPLIT_TOKEN);
        sb.append(_imageScaleFactor).append(SPLIT_TOKEN);
        sb.append(_initImages).append(SPLIT_TOKEN);
        sb.append(_plotState).append(SPLIT_TOKEN);
        sb.append(_desc).append(SPLIT_TOKEN);
        sb.append(_dataDesc).append(SPLIT_TOKEN);
        for(int i=0; (i<_fitsData.length);i++) {
            sb.append(_fitsData[i]);
            if (i<_fitsData.length-1) {
                sb.append(SPLIT_TOKEN);
            }

        }
        return sb.toString();
    }

    public static WebPlotInitializer parse(String s) {
        if (s==null) return null;
        String sAry[]= s.split(SPLIT_TOKEN,13);
        WebPlotInitializer retval= null;
        if (sAry.length>=10 && sAry.length<=12) {
            try {
                int i= 0;
                CoordinateSys imageCoordSys= CoordinateSys.parse(sAry[i++]);
                Projection    projection= ProjectionSerializer.deserializeProjection(sAry[i++]);
                int           dataWidth= Integer.parseInt(sAry[i++]);
                int           dataHeight= Integer.parseInt(sAry[i++]);
                int           imageScaleFactor= Integer.parseInt(sAry[i++]);
                PlotImages    initImages= PlotImages.parse(sAry[i++]);
                PlotState     plotState= PlotState.parse(sAry[i++]);
                String        desc= getString(sAry[i++]);
                String        dataDesc= getString(sAry[i++]);
                List<WebFitsData> fdList= new ArrayList<WebFitsData>(3);
                while (i<sAry.length) {
                   fdList.add(WebFitsData.parse(sAry[i++]));
                }
                WebFitsData fitsData[]= fdList.toArray(new WebFitsData[fdList.size()]);
                retval= new WebPlotInitializer(plotState,initImages,imageCoordSys,projection,dataWidth,dataHeight,
                                               imageScaleFactor,fitsData,desc,dataDesc);

            } catch (NumberFormatException e) {
                retval= null;
            }
        }
        return retval;

    }

    private static String getString(String s) { return s.equals("null") ? null : s; }
}

File: src/firefly/js/visualize/WebPlotInitializer.js
/*
 * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
 */


import PlotImages from './PlotImages.js';
import CoordinateSys from './CoordSys.js';
import PlotState from './PlotState.js';
import WebFitsData from './WebFitsData.js';
import join from 'underscore.string/join';
import {parseInt, checkNull,getStringWithNull } from '../util/StringUtils.js';


// todo: ProjectionSerializer
// todo: Projection

const SPLIT_TOKEN= '--WebPlotInitializer--';


class WebPlotInitializer {
    /**
     *
     * @param {PlotState} plotState
     * @param {PlotImages} images
     * @param {CoordinateSys} imageCoordSys
     * @param {Projection} projection, a projection object
     * @param {number} dataWidth
     * @param {number} dataHeight
     * @param {number} imageScaleFactor
     * @param {array} fitsData, an array by band of WebFitsData
     * @param {string} desc
     * @param {string} dataDesc
     */
    constructor(plotState, images, imageCoordSys, projection, dataWidth, dataHeight,
                imageScaleFactor, fitsData, desc, dataDesc) {

        this.plotState= plotState;
        this.initImages= images;
        this.imageCoordSys= imageCoordSys;
        this.projectionSerialized= ProjectionSerializer.serializeProjection(projection);
        this.dataWidth= dataWidth;
        this.dataHeight= dataHeight;
        this.imageScaleFactor= imageScaleFactor;
        this.fitsData= fitsData;
        this.desc= desc;
        this.dataDesc= dataDesc;
    }

//======================================================================
//----------------------- Public Methods -------------------------------
//======================================================================

    /**
     *
     * @return {PlotState}
     */
    getPlotState() { return this.plotState; }


    /**
     * @return {CoordinateSys}
     */
    getCoordinatesOfPlot() { return this.imageCoordSys; }


    /**
     * @return {*}
     */
    getInitImages() { return this.initImages; }

    /**
     *
     * @return {Projection}
     */
    getProjection() {
        return ProjectionSerializer.deserializeProjection(this.projectionSerialized);
    }

    /**
     *
     * @return {number|*}
     */
    getDataWidth() { return this.dataWidth; }
    /**
     *
     * @return {number|*}
     */
    getDataHeight() { return this.dataHeight; }
    /**
     *
     * @return {number|*}
     */
    getImageScaleFactor() { return this.imageScaleFactor; }


    /**
     *
     * @return {WebFitsData[]}
     */
    getFitsData()  { return this.fitsData; }

    /**
     * @return {string}
     */
    getPlotDesc() { return this.desc; }


    /**
     *
     * @param {string} d the description
     */
    setPlotDesc(d) { this.desc= d; }

    /**
     *
     * @return {string}
     */
    getDataDesc() { return this.dataDesc; }

    toString() {
        var part1= join(SPLIT_TOKEN,
            this.imageCoordSys, this.projectionSerialized,
            this.dataWidth, this.dataHeight,
            this.imageScaleFactor, this.initImages,
            this.plotState, this.desc,
            this.dataDesc);

        this.fitsData.length= 3;
        var part2= this.fitsData.map( (fd) => fd ? fd.toString() : 'null');

        return join(SPLIT_TOKEN, part1, part2);
    }

    static parse(s) {
        if (!s) return null;
        var sAry= s.split(SPLIT_TOKEN,13);
        if (sAry.length<10 || sAry.length>12) return null;

        var i= 0;
        var imageCoordSys= CoordinateSys.parse(sAry[i++]);
        var projection= ProjectionSerializer.deserializeProjection(sAry[i++]);
        var dataWidth= parseInt(sAry[i++]);
        var dataHeight= parseInt(sAry[i++]);
        var imageScaleFactor= parseInt(sAry[i++]);
        var initImages= PlotImages.parse(sAry[i++]);
        var plotState= PlotState.parse(sAry[i++]);
        var desc= getStringWithNull(sAry[i++]);
        var dataDesc= getStringWithNull(sAry[i++]);
        var fitsData= [undefined,undefined,undefined];
        while (i<sAry.length) {
            fitsData.push(WebFitsData.parse(sAry[i++]));
        }
        return new WebPlotInitializer(
            plotState,initImages,imageCoordSys,projection,
            dataWidth,dataHeight,
            imageScaleFactor,fitsData,desc,dataDesc);
    }

}

export default WebPlotInitializer;
